File: VB.Net/PIT_Magic/PitInputStream.vb
Imports Microsoft.VisualBasic.CompilerServices
Imports System
Imports System.IO
Imports System.Text

Namespace PIT_Magic
    Public Class PitInputStream
        ' Methods
        Public Sub New(ByVal byteStream As Stream)
            Me.inputStream = New BinaryReader(byteStream, Encoding.UTF8)
        End Sub

        Public Function GetMemoryAddress() As String
            Return $"0x{Me.inputStream.BaseStream.Position.ToString("X")}"
        End Function

        Public Function Read(ByVal buffer As Byte(), ByVal offset As Integer, ByVal length As Integer) As Integer
            Return Me.inputStream.Read(buffer, offset, length)
        End Function

        Public Function ReadBytes(ByVal byteCount As Integer) As Byte()
            Return Me.inputStream.ReadBytes(byteCount)
        End Function

        Public Function ReadInt() As Integer
            Return Me.inputStream.ReadInt32
        End Function

        Public Function ReadRemainingBytes() As String
            Dim count As Integer = CInt((Me.inputStream.BaseStream.Length - Me.inputStream.BaseStream.Position))
            Dim buffer As Byte() = New Byte(0  - 1) {}
            buffer = Me.inputStream.ReadBytes(count)
            Dim builder As New StringBuilder((buffer.Length * 2))
            Dim num2 As Byte
            For Each num2 In buffer
                builder.AppendFormat("{0:X2}", num2)
            Next
            Return builder.ToString
        End Function

        Public Function ReadShort() As Short
            Return Me.inputStream.ReadInt16
        End Function

        Public Function ReadString(ByVal buffer As Byte(), ByVal strMaxLength As Integer) As String
            Dim chArray As Char() = New Char((strMaxLength + 1)  - 1) {}
            If (Me.inputStream.Read(buffer, 0, strMaxLength) <> strMaxLength) Then
                Return Conversions.ToString(-1)
            End If
            Dim index As Integer = 0
            Dim i As Integer
            For i = 0 To strMaxLength - 1
                If (buffer(i) = 0) Then
                    index = i
                    Exit For
                End If
                chArray(i) = Convert.ToChar(buffer(i))
            Next i
            chArray(index) = Convert.ToChar(0)
            Return New String(chArray, 0, index)
        End Function


        ' Fields
        Private inputStream As BinaryReader
    End Class
End Namespace

File: C#/PIT_Magic/PitInputStream.cs
namespace PIT_Magic
{
    using Microsoft.VisualBasic.CompilerServices;
    using System;
    using System.IO;
    using System.Text;

    public class PitInputStream
    {
        private BinaryReader inputStream;

        public PitInputStream(Stream byteStream)
        {
            this.inputStream = new BinaryReader(byteStream, Encoding.UTF8);
        }

        public string GetMemoryAddress() => 
            $"0x{this.inputStream.BaseStream.Position.ToString("X")}";

        public int Read(byte[] buffer, int offset, int length) => 
            this.inputStream.Read(buffer, offset, length);

        public byte[] ReadBytes(int byteCount) => 
            this.inputStream.ReadBytes(byteCount);

        public int ReadInt() => 
            this.inputStream.ReadInt32();

        public string ReadRemainingBytes()
        {
            int count = (int) (this.inputStream.BaseStream.Length - this.inputStream.BaseStream.Position);
            byte[] buffer = new byte[0];
            buffer = this.inputStream.ReadBytes(count);
            StringBuilder builder = new StringBuilder(buffer.Length * 2);
            foreach (byte num2 in buffer)
            {
                builder.AppendFormat("{0:X2}", num2);
            }
            return builder.ToString();
        }

        public short ReadShort() => 
            this.inputStream.ReadInt16();

        public string ReadString(byte[] buffer, int strMaxLength)
        {
            char[] chArray = new char[strMaxLength + 1];
            if (this.inputStream.Read(buffer, 0, strMaxLength) != strMaxLength)
            {
                return Conversions.ToString(-1);
            }
            int index = 0;
            for (int i = 0; i < strMaxLength; i++)
            {
                if (buffer[i] == 0)
                {
                    index = i;
                    break;
                }
                chArray[i] = Convert.ToChar(buffer[i]);
            }
            chArray[index] = Convert.ToChar(0);
            return new string(chArray, 0, index);
        }
    }
}

